src/builtin_string.cpp

```diff
@@ -58,31 +59,33 @@ static bool string_args_from_stdin(const io_streams_t &streams) {
 }
 
 static const wchar_t *string_get_arg_stdin(wcstring *storage, const io_streams_t &streams) {
-    std::string arg;
-    for (;;) {
-        char ch = '\0';
-        long rc = read_blocked(streams.stdin_fd, &ch, 1);
-
-        if (rc < 0) {  // failure
-            return NULL;
-        }
-
-        if (rc == 0) {  // EOF
-            if (arg.empty()) {
-                return NULL;
-            }
-            break;
+    // We might read more than a line - store the rest in a static buffer.
+    static std::string buffer;
+
+    // Read in chunks from fd until buffer has a line.
+    std::string::iterator pos;
+    while ((pos = std::find (buffer.begin(), buffer.end(), '\n')) == buffer.end ()) {
```

Comment:
This is a bit more idiomatic as:
```
    size_t pos;
    while ((pos = buffer.find('\n')) == std::string::npos) {...
```

Reply:
THAT WAS IT! I've been thinking that there was some nicer way to to `find`, but I couldn't for the life of me come up with it.